lib/swagger/diff/diff.rb

```diff
@@ -6,49 +6,85 @@ def initialize(old, new)
         @old_specification = Swagger::Diff::Specification.new(old)
       end
 
+      def changes
+        @changes ||= {
+          new_endpoints: new_endpoints.to_a.sort,
+          removed_endpoints: missing_endpoints.to_a.sort,
+          new_request_params: new_or_changed_request_params,
+          removed_request_params: incompatible_request_params,
+          new_response_attributes: new_or_changed_response_attributes,
+          removed_response_attributes: incompatible_response_attributes
+        }
+      end
+
+      def changes_message
+        changed_endpoints_message + changed_params_message + changed_attrs_message
+      end
+
       def compatible?
         endpoints_compatible? && requests_compatible? && responses_compatible?
       end
 
       def incompatibilities
-        { endpoints: missing_endpoints.to_a.sort,
+        @incompatibilities ||= {
+          endpoints: missing_endpoints.to_a.sort,
           request_params: incompatible_request_params,
-          response_attributes: incompatible_response_attributes }
+          response_attributes: incompatible_response_attributes
+        }
       end
 
       def incompatibilities_message
         msg = ''
-        if incompatibilities[:endpoints]
-          msg += incompatibilities_message_endpoints(incompatibilities[:endpoints])
-        end
-        if incompatibilities[:request_params]
-          msg += incompatibilities_message_params(incompatibilities[:request_params])
-        end
+        msg += endpoints_message('missing', incompatibilities[:endpoints]) if incompatibilities[:endpoints]
+        msg += params_message('incompatible', incompatibilities[:request_params]) if incompatibilities[:request_params]
         if incompatibilities[:response_attributes]
-          msg += incompatibilities_message_attributes(incompatibilities[:response_attributes])
+          msg += attributes_message('incompatible', incompatibilities[:response_attributes])
         end
         msg
       end
 
       private
 
-      def incompatibilities_message_endpoints(endpoints)
+      def changed_endpoints_message
+        msg = ''
+        msg += endpoints_message('new', changes[:new_endpoints]) if changes[:new_endpoints]
```

Comment:
This looks like it will be easy to introduce a copy and paste bug. How about creating a message builder that only adds the message if there is a change?